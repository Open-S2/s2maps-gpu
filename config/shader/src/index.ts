/** @module Shader */
import GLSL from './glsl';
import WGSL from './wgsl';

export const GLSLLinker = GLSL;
export const WGSLLinker = WGSL;

export { getBundleHash, getBundleKey, getBundleEntry, toBundle, toModule } from './util';
export * from './types';
