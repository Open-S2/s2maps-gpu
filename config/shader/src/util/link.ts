import { Tree } from '@lezer/common';
import { ShaderModule, ParsedBundle, ParsedModule, ParsedModuleCache, ShaderDefine, ImportRef, RefFlags as RF } from '../types';
import { VIRTUAL_BINDINGS } from '../constants';

import { bindBundle, bindModule } from './bind';
import { toBundle, getBundleKey } from './bundle';
import { resolveShakeOps } from './shake';
import mapValues from 'lodash/mapValues';

export type Linker = (
  source: ParsedBundle,
  libraries?: Record<string, ShaderModule>,
) => string;

export type LoadModuleWithCache = (
  code: string,
  name: string,
  entry?: string,
  cache?: ParsedModuleCache | null,
) => ParsedModule;

export type GetPreambles = () => string[];

export type GetRenames = (
  defines?: Record<string, ShaderDefine> | null,
) => Map<string, string>;

export type DefineConstants = (
  defines: Record<string, ShaderDefine>
) => string;

export type DefineEnables = (
  enabled: string[],
) => string;

export type RewriteUsingAST = (
  code: string,
  tree: Tree,
  rename: Map<string, string>,
  shake?: number[] | null,
  optionals?: Set<string> | null,
) => string;

const NO_LIBS: Record<string, ShaderModule> = {};

// Link a source module with static modules and dynamic links.
export const makeLinkCode = (
  linker: Linker,
  loadModuleWithCache: LoadModuleWithCache,
  defaultCache: ParsedModuleCache,
) => (
  code: string,
  libraries: Record<string, string> = {},
  links?: Record<string, string | null> | null,
  defines?: Record<string, ShaderDefine> | null,
  cache: ParsedModuleCache | null = defaultCache,
) => {
  const main = loadModuleWithCache(code, 'main', undefined, cache);

  const parsedLibraries = mapValues(libraries, (code: string, name: string) => loadModuleWithCache(code, name, undefined, cache));
  const parsedLinks = mapValues(links, (code: string, name: string) =>
    (
      code != null
      ? loadModuleWithCache(code, name.split(':')[0], undefined, cache)
      : null
    ) as ShaderModule | null
  ) as any;

  const bundle = bindModule(main, parsedLinks, defines);
  return linker(bundle, parsedLibraries);
};

// Link a bundle of parsed module + libs, dynamic links
export const makeLinkBundle = (
  linker: Linker,
) => (
  source: ShaderModule,
  links?: Record<string, ShaderModule | null>,
  defines?: Record<string, ShaderDefine> | null,
) => {
  let bundle = toBundle(source);
  if (links || defines) bundle = bindBundle(bundle, links, defines);

  return linker(bundle);
};

// Link a bundle of parsed module + libs, dynamic links
export const makeLinkModule = (
  linker: Linker,
) => (
  source: ParsedModule,
  libraries: Record<string, ShaderModule> = NO_LIBS,
  links?: Record<string, ShaderModule | null>,
  defines?: Record<string, ShaderDefine> | null,
) => {
  let bundle = toBundle(source);
  if (links || defines) bundle = bindBundle(bundle, links, defines);

  return linker(bundle, libraries);
};

// Make a shader linker with injectable language rules
export const makeLinker = (
  getPreambles: GetPreambles,
  getRenames: GetRenames,
  defineConstants: DefineConstants,
  defineEnables: DefineEnables,
  rewriteUsingAST: RewriteUsingAST,
) => (
  source: ShaderModule,
  libraries: Record<string, ShaderModule> = NO_LIBS,
) => {
  const bundle = toBundle(source);
  const main = getBundleKey(source);

  const {bundles, exported, imported, aliased} = loadBundlesInOrder(bundle, libraries);
  const program = getPreambles();

  // Gather defines/enables
  const defs: Record<string, ShaderDefine> = {};
  const enabled: string[] = [];
  for (const {defines, module: {table: {enables}}} of bundles) {
    if (enables) for (let k of enables) enabled.push(k);
    if (defines) for (let k in defines) defs[k] = defines[k];
  }

  const en = defineEnables(enabled)
  if (en.length) program.push(en);

  const staticRename = getRenames(defs);
  const def = defineConstants(defs);
  if (def.length) program.push(def);

  // Namespace by module key
  const namespaces = new Map<number, string>();

  // Track symbols in global namespace 
  const exists = new Set<string>();
  const visible = new Set<string>();
  const fixed = new Map<string, string>();

  // Track link signatures
  const signatures = new Map<string, any>();
  const infers = new Map<string, string>();

  let hasBoundVirtuals = false;

  for (const bundle of bundles) {
    const {module, defines} = bundle;
    const {name, code, tree, table, shake, virtual} = module;
    const {globals, symbols, visibles, externals, modules, exports: exp} = table;

    const key = getBundleKey(bundle);
    const importMap = imported.get(key);
    const aliasMap = aliased.get(key);

    let optionals: Set<string> | null = null;

    // Namespace all non-global symbols outside main module
    let scope = '';
    const rename = new Map<string, string>();
    if (key !== main) {
      const namespace = virtual?.namespace;
      const ns = reserveNamespace(key, namespaces, namespace);
      scope = ns;

      if (symbols) for (const name of symbols) rename.set(name, ns + name);
      if (globals) for (const name of globals) {
        rename.set(name, name);
        fixed.set(ns + name, name);
      }
      if (visibles) for (const name of visibles) visible.add(rename.get(name)!);
      for (const name of rename.values()) exists.add(name);

      // Gather all exported signatures for type inference
      if (exp) for (const {flags, func} of exp) {
        if (func && (flags & RF.Exported)) {
          const {name} = func;
          signatures.set(ns + name, func);
        }
      }
    }

    // Replace imported symbol names with target
    if (modules) for (const {name: module, imports} of modules) {
      const key = importMap!.get(module)!;
      const ns = namespaces.get(key);

      for (const {name, imported} of imports) {
        let imp = ns + imported;
        if (fixed.has(imp)) imp = fixed.get(imp)!;
        if (!exists.has(imp)) {
          console.warn(`Import ${name} from '${module}' does not exist`);
          debugger;
        }
        else if (!visible.has(imp)) console.warn(`Import ${name} from '${module}' is private`);
        rename.set(name, imp);
        infers.set(scope + name, imp);
      }
    }

    // Replace imported function prototype names with target
    if (externals) for (const {flags, func, variable, struct} of externals) if (func ?? variable ?? struct) {
      const {name, inferred} = func ?? variable ?? struct;
      const key = importMap?.get(name)!;
      const ns = namespaces.get(key);

      const resolved = aliasMap?.get(name) ?? name;
      if ((ns === undefined) && (flags & RF.Optional)) {
        if (!optionals) optionals = new Set();
        optionals!.add(name);
        continue;
      }

      let imp = ns + resolved;
      if (fixed.has(imp)) imp = fixed.get(imp)!;
      if (!exists.has(imp)) {
        console.warn(`Link ${name}:${resolved} does not exist`);
        debugger;
      }
      else if (!visible.has(imp)) console.warn(`Link ${name}:${resolved} is private`);
      rename.set(name, imp);

      if (inferred) {
        const sig = signatures.get(imp);
        const {type, parameters} = sig;
        for (const {name, at} of inferred) {
          const resolved = at < 0 ? type : parameters[at];

          let imp = ns + (resolved.type ?? resolved.name ?? resolved);
          let i = imp;
          while (i = infers.get(imp)) { imp = i; }

          rename.set(name, imp);
          infers.set(scope + name, imp);
        }
      }
    }

    // Copy over static renames
    for (let k of staticRename.keys()) rename.set(k, staticRename.get(k)!);

    if (name === VIRTUAL_BINDINGS) hasBoundVirtuals = true;

    if (virtual) {
      const {uniforms, storages, textures} = virtual;
      if ((uniforms || storages || textures) && (!hasBoundVirtuals)) {
        const id = code.replace('@virtual ', '');
        throw new Error(`Virtual module ${id} has unresolved data bindings`);
      }

      // Emit virtual module in target namespace,
      // with dynamically assigned binding slots.
      const ns = namespaces.get(key)!;
      const recode = virtual.render(ns, rename, virtual.bindingBase, virtual.volatileBase);
      program.push(recode);
    }
    else if (tree) {
      // Shake tree ops based on which symbols were exported
      const keep = exported.get(key);
      const ops = shake && keep ? resolveShakeOps(shake, keep, symbols) : null;

      // Rename symbols using AST while tree shaking
      const recode = rewriteUsingAST(code, tree, rename, ops, optionals);
      program.push(recode);
    }
    else {
      // Static include
      program.push(code);
    }
  }

  const code = program.join("\n");
  return code;
};

// Load all references from a tree of bundles
// while gathering info about what's exported (for tree shaking).
export const loadBundlesInOrder = (
  bundle: ParsedBundle,
  libraries: Record<string, ShaderModule> = {},
): {
  bundles: ParsedBundle[]
  exported: Map<number, Set<string>>,
  imported: Map<number, Map<string, number>>,
  aliased: Map<number, Map<string, string>>,
} => {
  const graph = new Map<number, number[]>();
  const seen  = new Set<number>();
  const hoist = new Set<number>();

  const exported = new Map<number, Set<string>>();
  const imported = new Map<number, Map<string, number>>();
  const aliased  = new Map<number, Map<string, string>>();

  const out: ParsedBundle[] = [];

  const {module} = bundle;
  const {name, entry} = module;
  const key = getBundleKey(bundle);
  exported.set(key, new Set([entry ?? 'main']));

  // Traverse graph starting from source
  const queue = [{key, name, chunk: bundle as ShaderModule}];
  seen.add(key);

  const getContext = (m: ParsedModule) => {
    const {name, code} = m;
    if (name.match(/^_[A-Z]{2}_/) && code.match(/^@/)) return code.replace(/^@/, '');
    return name;
  }

  while (queue.length) {
    const next = queue.shift()!;
    const {key, name, chunk} = next;
    if (chunk == null) throw new Error(`Module '${name}' is undefined`);

    const bundle = toBundle(chunk);
    const {module, libs, links: linkDefs} = bundle;
    const {table: {modules, externals}} = module;
    const deps = [] as number[];

    const [links, aliases] = parseLinkAliases(linkDefs);

    // Static renames and imports for this module instance
    let aliasMap: Map<string, string> | null = null;
    let importMap: Map<string, number> | null = null;

    // Recurse into imports
    if (modules) for (const {at, name, imports} of modules) {
      const chunk = libs?.[name] ?? libraries[name];
      if (!chunk) throw new Error(`Module '${name}' in ${getContext(module)} is unknown`);

      const key = getBundleKey(chunk);
      if (!seen.has(key)) queue.push({key, name, chunk});
      seen.add(key);
      deps.push(key);

      if (!importMap) importMap = new Map();
      importMap!.set(name, key);

      if (at < 0) hoist.add(key);

      let list = exported.get(key);
      if (!list) exported.set(key, list = new Set());
      imports.forEach((i: ImportRef) => list!.add(i.imported));
    }

    // Recurse into links
    if (externals) for (const {flags, func, variable, struct} of externals) if (func ?? variable ?? struct) {
      const {name} = func ?? variable ?? struct;
      const chunk = links[name];
      if (!chunk) {
        if (flags & RF.Optional) {
          continue;
        }
        throw new Error(`Link '${name}' in ${getContext(module)} is not linked`);
      }

      const key = getBundleKey(chunk);
      if (!seen.has(key)) queue.push({key, name, chunk});
      seen.add(key);
      deps.push(key);

      if (!importMap) importMap = new Map();
      importMap!.set(name, key);

      const entry = chunk.entry ?? (chunk as any).module?.entry;
      const symbol = entry ?? aliases?.get(name) ?? name;

      if (symbol !== name) {
        if (!aliasMap) aliasMap = new Map();
        aliasMap!.set(name, symbol);
      }

      let list = exported.get(key);
      if (!list) exported.set(key, list = new Set());
      list.add(symbol);
    }

    // Build module-to-module dependency graph
    graph.set(key, deps);

    // Store aliases/imports/bundle
    if (aliasMap) aliased.set(key, aliasMap);
    if (importMap) imported.set(key, importMap);

    out.push(bundle);
  }

  // Sort by graph depth
  const order = getGraphOrder(graph, key);
  for (let [k, v] of order.entries()) if (hoist.has(k)) order.set(k, v + 1e5);
  out.sort((a, b) => order.get(getBundleKey(b))! - order.get(getBundleKey(a))! || a.module.name.localeCompare(b.module.name));

  return {
    bundles: out,
    exported,
    imported,
    aliased,
  };
};


// Generate a new namespace
export const reserveNamespace = (
  key: string | number,
  namespaces: Map<any, string>,
  force?: string,
): string => {
  let namespace = force ?? '_' + ('00' + (namespaces.size + 1).toString(36)).slice(-2) + '_';
  namespaces.set(key, namespace);
  return namespace;
}

// Get depth for each item in a graph, so its dependencies resolve correctly
export const getGraphOrder = (
  graph: Map<number, number[]>,
  name: number,
  depth: number = 0,
) => {
  const queue = [{name, depth: 0, path: [name]}];
  const depths = new Map<number, number>();

  while (queue.length) {
    const {name, depth, path} = queue.shift()!;
    depths.set(name, depth);

    const module = graph.get(name);
    if (!module) continue;

    const deps = module.map(name => {
      const i = path.indexOf(name);
      if (i >= 0) throw new Error("Cycle detected in module dependency graph: " + path.slice(i));
      return {name, depth: depth + 1, path: [...path, name]};
    });

    queue.push(...deps);
  }
  return depths;
}

// Parse run-time specified keys `from:to` into a map of aliases
export const parseLinkAliases = <T>(
  links?: Record<string, T>,
): [
  Record<string, T>,
  Map<string, string> | null,
] => {
  const out = {} as Record<string, T>;
  let aliases = null as Map<string, string> | null;

  for (let k in links) {
    const link = links[k] as any;
    if (!link) continue;

    let [name, imported] = k.split(':');
    if (!imported && link.entry != null) imported = link.entry;
    if (!imported && link.module?.entry != null) imported = link.module.entry;

    out[name] = link;
    if (imported) {
      if (!aliases) aliases = new Map<string, string>();
      aliases.set(name, imported);
    }
  }

  return [out, aliases];
}
