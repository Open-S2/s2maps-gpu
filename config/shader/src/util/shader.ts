import { Tree } from '@lezer/common';
import { ASTParser, VirtualTable, SymbolTableT, ParsedBundle, ParsedModule, ParsedModuleCache, CompressedNode } from '../types';
import { formatMurmur53, toMurmur53 } from './hash';
import { PREFIX_VIRTUAL } from '../constants';

const EMPTY_LIST = [] as any[];
const EMPTY_TABLE = {} as any;

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const makeLoadModule = <T extends SymbolTableT = any>(
  parseShader: (code: string) => Tree,
  makeASTParser: (code: string, tree: Tree, name?: string) => ASTParser<T>,
  compressAST: (code: string, tree: Tree, symbols?: string[]) => CompressedNode[],
  decompressAST: (nodes: CompressedNode[], symbols?: string[]) => Tree,
) => (
  code: string,
  name: string = 'main',
  entry?: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  if (typeof code !== 'string') throw new Error(`Shader code ${name} is not a string`);
  let tree = parseShader(code);

  const astParser = makeASTParser(code, tree, name);
  const table = astParser.getSymbolTable();
  const shake = astParser.getShakeTable(table);

  if (compressed) {
    const {symbols} = table;
    tree = decompressAST(compressAST(code, tree, symbols), symbols);
  }
  const hash = toMurmur53(code);

  return bindEntryPoint({name, code, hash, table, shake, tree}, entry);
}

// Use cache to load modules
export const makeLoadModuleWithCache = (
  loadModule: (code: string, name?: string, entry?: string, compressed?: boolean) => ParsedModule,
  defaultCache: ParsedModuleCache,
) => (
  code: string,
  name?: string,
  entry?: string,
  cache: ParsedModuleCache | null = defaultCache,
): ParsedModule => {
  if (!cache) return loadModule(code, name, entry, true);

  const hash = toMurmur53(code);
  const cached = cache.get(hash);
  if (cached) {
    return bindEntryPoint(cached, entry);
  }
  
  const module = loadModule(code, name, undefined, true);
  cache.set(hash, module);
  return bindEntryPoint(module, entry);
}

// Load a static (inert) module
export const loadStaticModule = (code: string, name: string, entry?: string) => {
  const hash = toMurmur53([code, entry]);
  return ({ name, code, hash, table: EMPTY_TABLE });
}

// Load a virtual (generated) module
export const loadVirtualModule = <T extends SymbolTableT = any>(
  virtual: VirtualTable,
  initTable: Partial<T> = EMPTY_TABLE,
  entry?: string,
  hash?: number,
  code?: string,
  key?: number,
) => {
  let symbols = initTable.symbols ?? EMPTY_LIST;

  code = code ?? `@virtual [${symbols.join(' ')}]`;
  hash = hash ?? toMurmur53(code);
  key  = key  ?? hash;

  const name = `${PREFIX_VIRTUAL}${formatMurmur53(key).slice(0, 6)}`;

  const table = {
    symbols,
    visibles: symbols,
    ...initTable,
  };
  return { name, code, hash, table, entry, virtual, key };
}

// Set entry point of a module, returns new bundle/module.
// Is the same instance as the original (key = old key/hash), so it merges with copies of itself.
// But is structurally different (hash = new key), so differences in links are reflected in the shader hash.
export const bindEntryPoint = <T extends ParsedBundle | ParsedModule>(bundle: T, entry?: string): T => {
  let {key, hash, module, table} = bundle as any;

  table = table ?? module?.table;
  hash = hash ?? module?.hash;

  if (entry == null && table.symbols?.includes('main')) entry = 'main';
  if (entry == null) return bundle;

  const structural = toMurmur53([hash, entry]);
  return {...bundle, entry, hash: structural, key: key ?? hash};
};
