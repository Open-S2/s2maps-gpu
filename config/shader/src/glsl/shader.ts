import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse GLSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make GLSL definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

/** Make GLSL enable definitions */
export const defineEnables = (enabled: string[]) => enabled.map(e => `#extension ${e}â€‹ : enable`).join('\n');

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace removal
const minifyCode = (code: string) => {
  code = code.replace(/ +/g, ' ');
  code = code.replace(/\n+/g, '\n');
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileGLSL = makeTranspile('glsl', 'glsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
glsl`...`
``` */
export const glsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as GLSL `float` */
export const float = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : 0;  
};
/** Format `number` as GLSL `uint` */
export const uint = (x: number) => Math.round(x).toString();
/** Format `number` as GLSL `int` */
export const int = (x: number) => Math.round(x).toString();
