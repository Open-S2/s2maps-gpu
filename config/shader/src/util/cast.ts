import { ShaderModule, ParsedBundle, UniformAttribute, RefFlags as RF } from '../types';
import { loadVirtualModule } from './shader';
import { toMurmur53, scrambleBits53, mixBits53 } from './hash';
import { toBundle, getBundleHash, getBundleKey } from './bundle';
import { PREFIX_CAST } from '../constants';

const NO_SYMBOLS = [] as string[];

export type CastTo = {
  basis: string,
  signs?: string,
  gain?: number,
};

export type BundleToAttribute = (
  bundle: ShaderModule,
) => UniformAttribute;

export type MakeCastAccessor = (
  name: string,
  accessor: string,
  args: string[],
  from: string,
  to: string,
  swizzle: string | CastTo,
) => string;

export type MakeSwizzleAccessor = (
  name: string,
  from: string,
  to: string,
  swizzle: string | CastTo,
) => string;

const SWIZZLE_SYMBOLS = ['swizzle'];

const CAST_SYMBOLS = ['cast', 'getValue'];
const CAST_EXTERNALS = [{
  func: {name: 'getValue'},
  flags: RF.External,
}];

const makeDeclarations = (name: string, type: any, parameters: any) => [{
  func: {name, type, parameters},
  flags: RF.Exported,
}] as any[];

export const makeSwizzleTo = (
  makeSwizzleAccessor: MakeSwizzleAccessor,
) => (
  from: string,
  to: string,
  swizzle: string | CastTo,
): ParsedBundle => {
  const entry = 'swizzle';

  const id = mixBits53(toMurmur53(swizzle), mixBits53(toMurmur53(from), toMurmur53(to)));

  const code = `@swizzle`;
  const hash = scrambleBits53(mixBits53(toMurmur53(code), id));
  const key  = hash;

  // Code generator
  const render = (namespace: string, rename: Map<string, string>) => {
    const name = rename.get(entry) ?? entry;
    return makeSwizzleAccessor(name, from, to, swizzle);
  }

  const exports = makeDeclarations(entry, to, [from]);

  const module = loadVirtualModule(
    { render },
    { symbols: SWIZZLE_SYMBOLS, exports },
    entry,
    hash,
    code,
    key,
  );

  return {module};
}

export const makeCastTo = (
  makeCastAccessor: MakeCastAccessor,
  bundleToAttribute: BundleToAttribute,
) => (
  source: ShaderModule,
  type: string,
  swizzle: string | CastTo,
): ParsedBundle => {
  const bundle = toBundle(source);

  const {module, virtuals} = bundle;
  const {name, format, args} = bundleToAttribute(bundle);

  const entry = 'cast';

  const hash = getBundleHash(bundle);
  const key  = getBundleKey(bundle);

  const id     = toMurmur53(swizzle);
  const code   = `@cast [${name} ${format}]`;
  const rehash = scrambleBits53(mixBits53(toMurmur53(code), mixBits53(hash, id)));
  const rekey  = scrambleBits53(mixBits53(rehash, key));

  // Code generator
  const render = (namespace: string, rename: Map<string, string>) => {
    const name = rename.get(entry) ?? entry;
    const accessor = rename.get('getValue') ?? 'getValue';
    return makeCastAccessor(name, accessor, args ?? [], format, type, swizzle);
  }

  const exports = makeDeclarations(entry, type, args);

  const cast = loadVirtualModule(
    { render },
    { symbols: CAST_SYMBOLS, exports, externals: CAST_EXTERNALS },
    entry,
    rehash,
    code,
    rekey,
  );

  const revirtuals = module.virtual
    ? (virtuals ? [...virtuals, module] : [module])
    : virtuals;

  return {
    module: cast,
    links: {
      getValue: bundle,
    },
    virtuals: revirtuals,
  };
}

export const parseSwizzle = (swizzle: string | CastTo) => {
  let c: CastTo;
  if (typeof swizzle === 'string') c = {basis: swizzle};
  else c = swizzle as CastTo;

  let {basis} = c;
  while (basis.length < 4) basis = basis + '0';

  return c;
}
