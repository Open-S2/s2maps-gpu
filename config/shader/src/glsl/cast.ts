import { makeCastTo, makeSwizzleTo, parseSwizzle, CastTo } from '../util/cast';
import { bundleToAttribute } from './shader';

const arg = (x: number) => String.fromCharCode(97 + x);

const literal = (v: number, neg: boolean, isFloat: boolean) => {
  if (neg) v = -v;
  if (isFloat) {
    let s = v.toString();
    if (!s.match(/[.eE]/)) s = s + '.0';
    return s;
  }
  return Math.round(v).toString();
};

export const makeAutoSwizzle = (from: string, to: string) => {
  const mf = from.match(/vec([0-9])/);
  const mt = to.match(/vec([0-9])/);
  
  const nf = mf ? +mf[1] : 1;
  const nt = mt ? +mt[1] : 1;

  const prefix = 'xyzw'.slice(0, Math.min(nf, nt));
  const suffix = '0001'.slice(prefix.length, nt);
  return prefix + suffix;
};

export const makeCastAccessor = (
  name: string,
  accessor: string,
  args: string[],
  from: string,
  to: string,
  swizzle: string | CastTo,
) => {
  const symbols = args.map((t, i) => `${arg(i)}`);

  const ret = makeSwizzle(from, to, 'v', swizzle);

  return `${to} ${name}(${symbols.map((s, i) => `${args[i]} ${s}`).join(', ')}) {
  ${to} v = ${accessor}(${symbols.join(', ')});
  return ${ret};
}
`;
}

export const makeSwizzleAccessor = (
  name: string,
  from: string,
  to: string,
  swizzle: string | CastTo,
) => {
  const ret = makeSwizzle(from, to, arg(0), swizzle);
  return `${to} ${name}(${from} ${arg(0)}) {
  return ${ret};
}
`;
}

export const makeSwizzle = (
  from: string,
  to: string,
  name: string = 'v',
  swizzle: string | CastTo | null = null,
) => {
  const sz = (swizzle == null) ? makeAutoSwizzle(from, to) : swizzle;

  const isFloat = !!from.match(/^(float|vec2|vec3|vec4|double|dvec2|dvec3|dvec4)$/);

  if (typeof sz === 'string' && sz.match(/^[xyzw]+$/)) {
    return name + '.' + sz;
  }

  const {basis, signs, gain} = parseSwizzle(sz);
  const out: string[] = basis.split('').map((v, i) => {
    const neg = !!(signs && signs[i] === '-');
    
    if (v.match(/[0-9]/)) return literal(+v, neg, isFloat);
    else return (neg ? '-' : '') + `${name}.${v}`;
  });
  
  const compact: string[] = [];
  let compat = true;
  for (let c of out) {
    const simple = !!c.match(/^[^.-]+\.[xyzw]$/);
    if (simple && compat && compact.length) {
      compact[compact.length - 1] += c.split('.')[1];
    }
    else {
      compact.push(c);
      compat = simple;
    }
  }

  let ret = `${to}(${compact.join(', ')})`;
  if (gain != null) ret = `${ret} * ${literal(gain, false, isFloat)}`;

  return ret;
}

export const castTo = makeCastTo(makeCastAccessor, bundleToAttribute);
export const swizzleTo = makeSwizzleTo(makeSwizzleAccessor);
