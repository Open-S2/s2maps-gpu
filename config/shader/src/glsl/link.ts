import { Tree } from '@lezer/common';

import { defineConstants, defineEnables, loadModule, loadModuleWithCache, DEFAULT_CACHE } from './shader';
import { rewriteUsingAST } from './ast';
import { makeLinker, makeLinkCode, makeLinkBundle, makeLinkModule } from '../util/link';

import { GLSL_VERSION } from './constants';

// Override GLSL version/prefix
let PREAMBLE = `#version ${GLSL_VERSION}`;
export const setPreamble = (s: string): string => PREAMBLE = s;
export const getPreamble = (): string => PREAMBLE;
const getPreambles = (): string[] => [getPreamble()];

// No preprocessor renames
const getRenames = () => new Map<string, string>();

export const linker     = makeLinker(getPreambles, getRenames, defineConstants, defineEnables, rewriteUsingAST);
export const linkBundle = makeLinkBundle(linker);
export const linkModule = makeLinkModule(linker);
export const linkCode   = makeLinkCode(linker, loadModuleWithCache, DEFAULT_CACHE);
