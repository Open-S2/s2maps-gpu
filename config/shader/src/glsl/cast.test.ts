import { loadModule } from './shader';
import { linkBundle } from './link';
import { castTo, swizzleTo } from './cast';

describe('cast', () => {
  
  it('swizzles', () => {

    const result = swizzleTo('vec3', 'vec3', 'xyx');    
    const recode = linkBundle(result);
    expect(recode).toMatchSnapshot();

  });
  
  it('casts', () => {
    
    const code = `
    #pragma export
    vec2 getValue() { return vec2(1.0, 2.0); }
    `;

    const mod = loadModule(code, 'code', 'getValue');
    const result = castTo(mod, 'vec3', 'xyx');
    
    const recode = linkBundle(result);
    expect(recode).toMatchSnapshot();

  });

  it('casts complex swizzle', () => {
    
    const code = `
      #pragma export
      vec2 getValue() { return vec2(1.0, 2.0); }
    `;

    const mod = loadModule(code, 'code', 'getValue');
    const result = castTo(mod, 'vec4<f32>', {
      basis: 'xy10',
      signs: '-+-+',
      gain: 2,
    });
    
    const recode = linkBundle(result);
    expect(recode).toMatchSnapshot();

  });

  it('casts compact swizzle', () => {
    
    const code = `
      #pragma export
      vec2 getValue() { return vec2(1.0, 2.0); }
    `;

    const mod = loadModule(code, 'code', 'getValue');
    const result = castTo(mod, 'vec4<f32>', {
      basis: 'xyzw',
      signs: '++-+',
      gain: 2,
    });
    
    const recode = linkBundle(result);
    expect(recode).toMatchSnapshot();

  });

  it('casts and links', () => {
    
    const code = `
    #pragma export
    vec2 getValue() { return vec2(1.0, 2.0); }
    `;

    const main = `
    vec3 getValue();
    void main() { getValue(); }
    `;

    const sub = loadModule(code, 'code', 'getValue');
    const mod = loadModule(main, 'main');
    const getValue = castTo(sub, 'vec3', 'xyx');

    const recode = linkBundle(mod, {getValue});
    expect(recode).toMatchSnapshot();

  });

});