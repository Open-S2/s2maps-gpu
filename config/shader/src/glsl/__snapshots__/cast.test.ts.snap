// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cast casts 1`] = `
"#version 450

    vec2 _01_getValue() { return vec2(1.0, 2.0); }
    
vec3 cast() {
  vec3 v = _01_getValue();
  return v.xyx;
}
"
`;

exports[`cast casts and links 1`] = `
"#version 450

    vec2 _01_getValue() { return vec2(1.0, 2.0); }
    
vec3 _02_cast() {
  vec3 v = _01_getValue();
  return v.xyx;
}


    vec3 _02_cast();
    void main() { _02_cast(); }
    "
`;

exports[`cast casts compact swizzle 1`] = `
"#version 450

      vec2 _01_getValue() { return vec2(1.0, 2.0); }
    
vec4<f32> cast() {
  vec4<f32> v = _01_getValue();
  return vec4<f32>(v.xy, -v.z, v.w) * 2.0;
}
"
`;

exports[`cast casts complex swizzle 1`] = `
"#version 450

      vec2 _01_getValue() { return vec2(1.0, 2.0); }
    
vec4<f32> cast() {
  vec4<f32> v = _01_getValue();
  return vec4<f32>(-v.x, v.y, -1.0, 0.0) * 2.0;
}
"
`;

exports[`cast swizzles 1`] = `
"#version 450
vec3 swizzle(vec3 a) {
  return a.xyx;
}
"
`;
